### PR TITLE
[Domain Management] Add Tracks

### LINKED ISSUE #21934: 🍎 Tracks Events
This task is to implement tracks events for the domain management flows.

We should be sure to include events / properties capable of tracking the number / proportion of domain counts for which the domain is not served by WPCOM, as this will be valuable to drive decisions on future designs.

Recheck that the new analytic events added make sense and document the changes in a new post or by enhancing the spec pc8HXX-1gR-p2

### PR DESCRIPTION
**Pull request by alpavanoglu**
Fixes #21934 

Refer to the specs and validate if the tracks are being fired.

## Regression Notes
1. Potential unintended areas of impact
N/A

2. What I did to test those areas of impact (or what existing automated tests I relied on)
N/A

3. What automated tests I added (or what prevented me from doing so)
N/A

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

Not related as it is a tracks PR.
UI Changes testing checklist:
- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] VoiceOver.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] iPhone and iPad. 
- [ ] Multi-tasking: Split view and Slide over. (iPad)

### CODE REVIEW

**Comment by mokagio** [comment]
I'm going to bump this to the next release because we'll be code freezing 23.8 today and this hasn't been approved yet.

**Comment by salimbraksa** [comment]
I tested this with the following steps:

1. Navigate to the "Me" tab
2. Tap "Domains"
   - [x] Expect `me_all_domains_tapped` to be tracked.
3. Expect the domains list to be shown
   - [x] Expect `all_domains_list_shown` to be tracked.
4. Tap a domain from the list
5. Expect the domain details webview to be shown
   - [ ] Expect `all_domains_domain_details_web_view_shown` to be tracked.
6. Go back to the all domains screen
7. Tap the "+" button
   - [x] Expect `all_domains_add_domain_tapped` to be tracked
8. Expect the domain search screen to be shown
   - [x] Expect `domains_dashboard_domains_search_shown <source: all_domains>` to be tracked.
9. Type a query to search for a domain
10. Tap a domain from the results list then tap **Select Domain**
    - [x] Expect `domains_dashboard_select_domain_tapped <source: all_domains>` to be tracked.
    - [x] Expect the tracked event to include the `domain_name` property with a value set as the picked domain
11. Expect the purchase domain screen to be shown
    - [x] Expect `domain_management_purchase_domain_screen_shown` to be tracked.
12. Tap "Get domain"
    - [x] Expect `domain_management_purchase_domain_get_domain_tapped` to be tracked.
13. Expect checkout webview to be shown
14. Go back to the purchase domains screen (dismiss the error if needed)
15. Tap "Choose Site"
    - [x] Expect `domain_management_purchase_domain_choose_site_tapped` to be tracked.
16. Choose a site
    - [x] Expect `site_switcher_domain_site_selected` to be tracked.
17. Complete the purchase (can use the store sandbox if needed)
    - [x] Expect `domain_management_purchase_domain_completed` to be tracked.
18. Repeat steps 1 - 9 (or tap back until you get back to the search screen)
19. Tap "Transer domain" button
    - [x] Expect `domains_dashboard_domains_search_transfer_domain_tapped` to be tracked
20. Navigate to the "My Site" tab
21. Tap "... More" button
22. Tap the "Domains" button (in the "Manage" section)
23. Tap the "ALL" button (on the top bar)
    - [x] Expect `domains_dashboard_all_domains_tapped` to be tracked

File: WordPress/Classes/ViewRelated/Domains/Domain registration/RegisterDomainSuggestions/RegisterDomainSuggestionsViewController.swift [hunk 1]
```diff
@@ -111,6 +111,14 @@ class RegisterDomainSuggestionsViewController: UIViewController {
...
         setupTransferFooterView()
     }
+   
+    override func viewDidAppear(_ animated: Bool) {
+        super.viewDidAppear(animated)
+        WPAnalytics.track(
+            .domainsDashboardDomainsSearchShown,
+            properties: [WPAppAnalyticsKeySource: "all_domains"]
+        )
+    }
```
Comment (salimbraksa) [inline WordPress/Classes/ViewRelated/Domains/Domain registration/RegisterDomainSuggestions/RegisterDomainSuggestionsViewController.swift:118]: [Suggestion] `viewDidAppear` can be called multiple times after a screen is loaded. For example, when pushing to a new screen then popping back to previous screen. What do you say we move this logic to `viewDidLoad`?

Comment (alpavanoglu) [reply WordPress/Classes/ViewRelated/Domains/Domain registration/RegisterDomainSuggestions/RegisterDomainSuggestionsViewController.swift:118]: I can do that. I think the shown events should trigger whenever the screen is shown but I'm not sure what we've been following in our codebase.

**Comment by salimbraksa** [comment]
Hey @alpavanoglu the event `domain_management_purchase_domain_site_selected` should be renamed to `site_switcher_domain_site_selected`.

The Site Switcher screen can be used in different contexts, for example, if you go to `My Site > Site Switcher` then select a site, the event `domain_management_purchase_domain_site_selected` will be fired, which is not correct in that case. 

We can add a `<source: all_domains>` parameter, but that's not a high priority if it's complex.

File: WordPress/Classes/Utility/Analytics/WPAnalyticsEvent.swift [hunk 2]
```diff
@@ -976,7 +976,7 @@ import Foundation
         case .domainsSearchRowSelected:
             return "domain_management_domains_search_row_selected"
         case .purchaseDomainSiteSelected:
```
Comment (salimbraksa) [inline WordPress/Classes/Utility/Analytics/WPAnalyticsEvent.swift]: Nit: @alpavanoglu Would be great to rename the enum too. What do you think?

**Review by salimbraksa — APPROVED** [review]
@alpavanoglu I've dropped 1 nit comment. But LGTM!

**Comment by hassaanelgarem** [comment]
@salimbraksa @alpavanoglu In the testing steps https://github.com/wordpress-mobile/WordPress-iOS/pull/22038#issuecomment-1827796333, I see an event `all_domains_domain_details_web_view_shown`, but I can't find it in the code. Am I missing something?

**Comment by salimbraksa** [comment]
@hassaanelgarem You're right, I'm not sure what exactly happened. Maybe I accidentally checked that item? I'll rerun the testing steps to double check other events are fired. But I just want point out that a similar event to `all_domains_domain_details_web_view_shown` is fired:
```
🔵 Tracked: webkitview_displayed <source: all-domains>
```

**Comment by hassaanelgarem** [comment]
> But I just want point out that a similar event to all_domains_domain_details_web_view_shown is fired:

@salimbraksa It is also fired during the purchase flow, so we can't depend on that.

**Comment by hassaanelgarem** [comment]
I also noticed that the `domains_dashboard_domains_search_shown` event is fired with `<source: all_domains>` regardless of the source.

**Comment by hassaanelgarem** [comment]
@salimbraksa @alpavanoglu Please perform a thorough test and report all existing issues. We should aim to have these resolved and merged into the 23.8 release branch.

**Comment by salimbraksa** [comment]
@hassaanelgarem I'll open a PR today resolving these issues. 🙏

**Pull request merged**